clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -7430,6 +7430,271 @@ static void handleSYCLIntelFPGAMaxConcurrencyAttr(Sema &S, Decl *D,
   S.AddSYCLIntelFPGAMaxConcurrencyAttr(D, A, E);
 }
 
+// Checks if an expression is a valid filter list for an add_ir_attributes_*
+// attribute. Returns true if an error occured.
+static bool checkAddIRAttributesFilterListExpr(Expr *FilterListArg, Sema &S,
+                                               const AttributeCommonInfo &CI) {
+  const auto *FilterListE = cast<InitListExpr>(FilterListArg);
+  for (const Expr *FilterElemE : FilterListE->inits())
+    if (!isa<StringLiteral>(FilterElemE))
+      return S.Diag(FilterElemE->getBeginLoc(),
+                    diag::err_sycl_add_ir_attribute_invalid_filter)
+             << CI;
+  return false;
+}
+
+// Returns true if a type is either an array of char or a pointer to char.
+static bool isAddIRAttributesValidStringType(QualType T) {
+  if (!T->isArrayType() && !T->isPointerType())
+    return false;
+  QualType ElemT = T->isArrayType()
+                       ? cast<ArrayType>(T.getTypePtr())->getElementType()
+                       : T->getPointeeType();
+  return ElemT.isConstQualified() && ElemT->isCharType();
+}
+
+// Checks if an expression is a valid attribute name for an add_ir_attributes_*
+// attribute. Returns true if an error occured.
+static bool checkAddIRAttributesNameExpr(Expr *NameArg, Sema &S,
+                                         const AttributeCommonInfo &CI) {
+  // Only strings and const char * are valid name arguments.
+  if (isAddIRAttributesValidStringType(NameArg->getType()))
+    return false;
+
+  return S.Diag(NameArg->getBeginLoc(),
+                diag::err_sycl_add_ir_attribute_invalid_name)
+         << CI;
+}
+
+// Checks if an expression is a valid attribute value for an add_ir_attributes_*
+// attribute. Returns true if an error occured.
+static bool checkAddIRAttributesValueExpr(Expr *ValArg, Sema &S,
+                                          const AttributeCommonInfo &CI) {
+  QualType ValType = ValArg->getType();
+  if (isAddIRAttributesValidStringType(ValType) || ValType->isNullPtrType() ||
+      ValType->isIntegerType() || ValType->isFloatingType() ||
```

Comment:
`isIntegerType()` already covers `isCharType()` and `isEnumeralType()` but I kind of wonder if the correct predicate we really want is `isAddIRAttributesValidStringType(ValType) || (ValType->isFundamentalType() && !ValType->isVoidType())`?

Reply:
I tried, and you're right that `isIntegerType()` considers those other two, but if I use `(ValType->isFundamentalType() && !ValType->isVoidType())` it will not accept enums, so it would need `isEnumeralType()` as well. So either I remove `isCharType()` and `isEnumeralType()` or I use

`isAddIRAttributesValidStringType(ValType) || ValType->isEnumeralType() || (ValType->isFundamentalType() && !ValType->isVoidType())`

Do you have a preference?

Reply:
Ah, good catch on the enumeration types; I thought it was allowing C-style enums but not scoped enumerations; I missed the `!EnumeralType()` in there.

I think you should remove `isCharType()` and `isEnumeralType()`; that'll match the intent with less words.

Reply:
Note that isIntegerType doesn't contain SCOPED enums, which I suspect you don't want either/anyway.  isFundamentalType also includes nullptr_t (and pointer types!), so perhaps we don't mean that.

I might switch to `ValType->isIntegralOrUnscopedEnumerationType() || <string thing> || FP>`

Reply:
It should accept `nullptr_t`, but it definitely shouldn't accept other pointer types than the ones we explicitly allow. `isIntegralOrUnscopedEnumerationType` works for me!

Reply:
Looking at [sycl_ext_oneapi_device_global](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_device_global.asciidoc) the properties actually use scoped enums. Granted we could either change the extensions or cast the property enums to integers when we generate the corresponding values for the attributes, but is there a reason we wouldn't want to allow scoped enums?

Reply:
Scoped enumerations are strongly typed and don't have an implicit conversion to an integer value: https://godbolt.org/z/GchM4bEdh. However, the waters are muddy with attribute arguments because they're kind of outside of many of the language rules. e.g., https://godbolt.org/z/WfWTGMTof (note how `@.args` has `112` in it).

Reply:
It also works as-is. I have changed the check to use `isIntegralOrEnumerationType()` and added test cases for scoped enums. Are you both okay with that?

Reply:
I'm fine with that.